### PR TITLE
Added detector library cases for terraform

### PR DESCRIPTION
**Pull request by braraiha**
- Added detector library cases for terraform.

- **Note:** All the test cases have 100% recall and precision.

### CODE REVIEW

File: terraform/src/detectors/terraform-exposed-lambda-env-secret/terraform-exposed-lambda-env-secret_compliant.tf [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-exposed-lambda-env-secret@v1.0 defects=0}
+resource "aws_lambda_function" "IdempotencyFunction" {
+  function_name = "IdempotencyFunction"
+  role          = aws_iam_role.IdempotencyFunctionRole.arn
+  runtime       = "python3.11"
+  handler       = "app.lambda_handler"
+  # Compliant: Hard-coded secrets does not exist in lambda environment.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-exposed-lambda-env-secret/terraform-exposed-lambda-env-secret_compliant.tf:10]: "No hardcoding in lambda environment"

File: terraform/src/detectors/terraform-implicit-eks-ssh-access/terraform-implicit-eks-ssh-access_compliant.tf [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-implicit-eks-ssh-access@v1.0 defects=0}
+resource "aws_eks_node_group" "private-node-group-1-tf" {
+  cluster_name  = aws_eks_cluster.eks-tf.name
+  node_group_name = format("%s-private-ng-1-%s", local.project_prefix, local.build_suffix)
+  node_role_arn  = aws_iam_role.workernodes.arn
+  subnet_ids =  [for i in aws_subnet.eks-internal: i.id]
+  instance_types = ["t3.small"]
+  # Compliant: `source_security_group_ids` is mentioned in here.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-implicit-eks-ssh-access/terraform-implicit-eks-ssh-access_compliant.tf:11]: "`source_security_group_ids` has been configured"

File: terraform/src/detectors/terraform-implicit-eks-ssh-access/terraform-implicit-eks-ssh-access_non-compliant.tf [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-implicit-eks-ssh-access@v1.0 defects=1}
+resource "aws_eks_node_group" "private-node-group-1-tf" {
+  cluster_name  = aws_eks_cluster.eks-tf.name
+  node_group_name = format("%s-private-ng-1-%s", local.project_prefix, local.build_suffix)
+  node_role_arn  = aws_iam_role.workernodes.arn
+  subnet_ids =  [for i in aws_subnet.eks-internal: i.id]
+  instance_types = ["t3.small"]
+  # Noncompliant: `source_security_group_ids` is not mentioned.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-implicit-eks-ssh-access/terraform-implicit-eks-ssh-access_non-compliant.tf:11]: "No source_security_group_ids has been configured"

File: terraform/src/detectors/terraform-restrict-assumed-role/terraform-restrict-assumed-role_compliant.tf [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-assumed-role@v1.0 defects=0}
+resource "aws_iam_role" "over-privilege-role2" {
+  name = "over-privilege-role"
+
+  # Compliant: Specific assume role policy principal is mentioned.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-assumed-role/terraform-restrict-assumed-role_compliant.tf:8]: "A specific assume role policy principal has been configured."

File: terraform/src/detectors/terraform-restrict-assumed-role/terraform-restrict-assumed-role_non-compliant.tf [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-assumed-role@v1.0 defects=1}
+resource "aws_iam_role" "over-privilege-role1" {
+  name = "over-privilege-role"
+
+  # Noncompliant: Specific assume role policy principal is not mentioned.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-assumed-role/terraform-restrict-assumed-role_non-compliant.tf:8]: "No specific assume role policy principal has been configured."

File: terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_compliant.tf [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-asterisk-iam-admin-policies@v1.0 defects=0}
+data "aws_iam_policy_document" "policy" {
+  version = "2012-10-17"
+
+  # Compliant: This IAM policy restricts administrative privileges.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_compliant.tf:8]: "IAM policy correctly restricts administrative privileges"

File: terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_non-compliant.tf [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-asterisk-iam-admin-policies@v1.0 defects=1}
+data "aws_iam_policy_document" "policy" {
+  version = "2012-10-17"
+
+  # Noncompliant: This IAM policy allows full "*-*" administrative privileges.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_non-compliant.tf:8]: "IAM policy grants full administrative privileges (*:*)"

File: terraform/src/detectors/terraform-restrict-eks-traffic-access/terraform-restrict-eks-traffic-access_non-compliant.tf [hunk 8]
```diff
@@ -0,0 +1,14 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-eks-traffic-access@v1.0 defects=1}
+resource "aws_eks_cluster" "test" {
+  name = "test"
+  role_arn = aws_iam_role.eksrole.arn
+  vpc_config {
+    subnet_ids = [aws_subnet.subnet1.id]
+    # Noncompliant: `endpoint_public_access` not disbled to avoid traffic.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-eks-traffic-access/terraform-restrict-eks-traffic-access_non-compliant.tf:10]: "`endpoint_public_access` is enabled, which allows public traffic"

File: terraform/src/detectors/terraform-restrict-iam-admin-policy/terraform-restrict-iam-admin-policy_compliant.tf [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-iam-admin-policy@v1.0 defects=0}
+resource "aws_iam_policy" "ec2_pricing" {
+  name        = "ec2_pricing"
+  description = "allow access to ec2 instance types and pricing information"
+  path        = "/"
+  policy = jsonencode({
+    Version = "2012-10-17"
+    # Compliant: IAM policies that allow full "*-*" administrative privileges is not created.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-iam-admin-policy/terraform-restrict-iam-admin-policy_compliant.tf:11]: "No full administrative privileges (*-*) are granted in the IAM policy"

File: terraform/src/detectors/terraform-restrict-iam-admin-policy/terraform-restrict-iam-admin-policy_non-compliant.tf [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-iam-admin-policy@v1.0 defects=1}
+resource "aws_iam_policy" "ec2_pricing" {
+  name        = "ec2_pricing"
+  description = "allow access to ec2 instance types and pricing information"
+  path        = "/"
+  policy = jsonencode({
+    Version = "2012-10-17"
+    # Noncompliant: IAM policies that allow full "*-*" administrative privileges is created.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-iam-admin-policy/terraform-restrict-iam-admin-policy_non-compliant.tf:11]: "The IAM policy grants full *-* administrative privileges"

File: terraform/src/detectors/terraform-restrict-iam-asterisk-action-critical/terraform-restrict-iam-asterisk-action-critical_compliant.tf [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+resource "aws_iam_policy" "LambdaDynamoDBPolicy" {
+  name        = "LambdaDynamoDBPolicy"
+  description = "IAM policy for Lambda function to access DynamoDB"
+  policy = jsonencode({
+    Version = "2012-10-17"
+    Statement = [
+      {
+        Sid    = "AllowDynamodbReadWrite"
+        Effect = "Allow"
+        # Compliant: IAM policy actions are specified.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-iam-asterisk-action-critical/terraform-restrict-iam-asterisk-action-critical_compliant.tf:14]: "IAM policy actions are specifically defined for DynamoDB operations"

File: terraform/src/detectors/terraform-restrict-iam-asterisk-action/terraform-restrict-iam-asterisk-action_compliant.tf [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-iam-asterisk-action@v1.0 defects=0}
+data "aws_iam_policy_document" "policy" {
+  version = "2012-10-17"
+
+  # Compliant: This IAM policy restricts administrative privileges.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-iam-asterisk-action/terraform-restrict-iam-asterisk-action_compliant.tf:8]: remove "This"

File: terraform/src/detectors/terraform-restrict-sqs-asterisk-action/terraform-restrict-sqs-asterisk-action_compliant.tf [hunk 13]
```diff
@@ -0,0 +1,33 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-sqs-asterisk-action@v1.0 defects=0}
+resource "aws_sqs_queue_policy" "allow_email_notification_events" {
+  count = local.emailing_enabled
+
+  queue_url = aws_sqs_queue.email_events[0].id
+
+  # Compliant: SQS policy have specific action.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-sqs-asterisk-action/terraform-restrict-sqs-asterisk-action_compliant.tf:10]: "The SQS policy specifies a particular action (SQS:SendMessage)."

File: terraform/src/detectors/terraform-restrict-sqs-asterisk-action/terraform-restrict-sqs-asterisk-action_non-compliant.tf [hunk 14]
```diff
@@ -0,0 +1,33 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-sqs-asterisk-action@v1.0 defects=1}
+resource "aws_sqs_queue_policy" "allow_email_notification_events" {
+  count = local.emailing_enabled
+
+  queue_url = aws_sqs_queue.email_events[0].id
+
+  # Noncompliant: SQS policy have ALL (*) actions.
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-sqs-asterisk-action/terraform-restrict-sqs-asterisk-action_non-compliant.tf:10]: "The SQS policy grants * (all) actions"

**Review by pratik-amazon — COMMENTED** [review]
Please address the suggested changes

File: terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_non-compliant.tf [hunk 15]
```diff
@@ -0,0 +1,17 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+
+# {fact rule=terraform-restrict-asterisk-iam-admin-policies@v1.0 defects=1}
+data "aws_iam_policy_document" "policy" {
+  version = "2012-10-17"
+
+  # Noncompliant: IAM policy grants full administrative privileges (:).
```
Comment (pratik-amazon) [inline terraform/src/detectors/terraform-restrict-asterisk-iam-admin-policies/terraform-restrict-asterisk-iam-admin-policies_non-compliant.tf:8]: I mistakenly added : at last please remove it.

**Review by pratik-amazon — COMMENTED** [review]
Please correct one change

**Review by pratik-amazon — COMMENTED** [review]
Minor modification, suggestion

**Review by pratik-amazon — APPROVED** [review]

**Review by atonaamz — APPROVED** [review]

**Pull request merged**